File: frontend/src/services/projects.ts
import api from "../api/client";
import type {
  Project,
  ProjectCreate,
  ProjectUpdate,
  ProjectWithTasks,
} from "../types";

export const getProjects = async (): Promise<Project[]> => {
  const response = await api.get<Project[]>("/projects");
  return response.data;
};

export const getProjectWithTasks = async (
  projectId: string,
): Promise<ProjectWithTasks> => {
  const response = await api.get<ProjectWithTasks>(`/projects/${projectId}`);
  return response.data;
};

export const createProject = async (
  project: ProjectCreate,
): Promise<Project> => {
  const response = await api.post<Project>("/projects", project);
  return response.data;
};

export const updateProject = async (
  projectId: string,
  project: ProjectUpdate,
): Promise<Project> => {
  const response = await api.put<Project>(`/projects/${projectId}`, project);
  return response.data;
};

export const deleteProject = async (projectId: string): Promise<void> => {
  await api.delete(`/projects/${projectId}`);
};

File: app/routers/projects.py
from fastapi import APIRouter, Depends, status
from sqlalchemy.orm import Session
from typing import List

from app.core.database import get_db
from app.schemas import project as schemas
from app.services import project_service

router = APIRouter(prefix="/projects", tags=["projects"])


@router.get("", response_model=List[schemas.Project])
def get_projects(
    skip: int = 0,
    limit: int = 100,
    db: Session = Depends(get_db)
):
    return project_service.get_all(db, skip=skip, limit=limit)


@router.get("/{project_id}", response_model=schemas.ProjectWithTasks)
def get_project(
    project_id: str,
    db: Session = Depends(get_db)
):
    return project_service.get_with_tasks(db, project_id=project_id)


@router.post("", response_model=schemas.Project, status_code=status.HTTP_201_CREATED)
def create_project(
    project: schemas.ProjectCreate,
    db: Session = Depends(get_db)
):
    return project_service.create(db, project_in=project)


@router.put("/{project_id}", response_model=schemas.Project)
def update_project(
    project_id: str,
    project: schemas.ProjectUpdate,
    db: Session = Depends(get_db)
):
    return project_service.update(db, project_id=project_id, project_in=project)


@router.delete("/{project_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_project(
    project_id: str,
    db: Session = Depends(get_db)
):
    project_service.delete(db, project_id=project_id)
    return None
